### ISSUE TITLE
How to explain the output of a network without target?

### ISSUE THREAD

**Issue by monk1337** [pytorch/captum#991]
## ❓ Questions and Help

Hi, I am trying to find ways to visualize(explain) the output of any pyg network. For example, I am taking a simple toy network using Graph conv:

```
import numpy as np
import torch

num_of_nodes = 4
num_of_feats = 5
output_dim   = 6

def get_data():
  adj         = torch.Tensor(np.random.randint(0,2,[num_of_nodes,num_of_nodes]))
  feats       = torch.Tensor(np.random.uniform(-1,1,[num_of_nodes, num_of_feats]))

  edge_indexs = adj.nonzero(as_tuple=False).t()
  return feats, edge_indexs
```

The network looks like this:

```
import torch
import torch.nn.functional as F
from torch_geometric.nn import GCNConv

class GCN(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = GCNConv(num_of_feats, output_dim)

    def forward(self, x, edge_index):
        x = self.conv1(x, edge_index)
        x = F.relu(x)
        return x
```

calling network:

```
feats, ed_in = get_data()
xg = GCN()(feats, ed_in)
xg
```

output:

```
tensor([[0.4645, 0.4537, 0.7191, 0.2981, 0.9345, 0.2394],
        [0.4699, 0.3964, 0.4689, 0.5432, 0.4923, 0.5697],
        [0.0000, 0.0000, 0.0019, 0.0000, 0.1044, 0.0000],
        [0.5781, 0.5397, 0.6135, 0.5073, 0.7209, 0.5636]],
       grad_fn=<ReluBackward0>)
```

Is there a way to visualize this output?

I am looking for a very simple example of how to visualize the output of the above network using captum?

What I have tried:

I tried to explain the embeddings :

```
from torch_geometric.utils.convert import from_networkx
from torch_geometric.nn import Explainer, GCNConv, to_captum
from captum.attr import IntegratedGradients

import matplotlib.pyplot as plt

import numpy as np
import networkx as nx


def numpy_to_graph(A, node_features=None):
    
    G = nx.from_numpy_array(A)
    if node_features != None:
        for n in G.nodes():
            for k,v in node_features.items():
                G.nodes[n][k] = v[n]
    return G


def get_data(adj, feats):
    G           = numpy_to_graph(adj, node_features={'x': feats})
    pyg_graph   = from_networkx(G)
    return pyg_graph


def captum_explain_graph(model, 
                         edge_index, 
                         adj_matrix, 
                         output_idx, 
                         device):

  data   = get_data(adj_matrix, edge_index)
  target = 0
  model  = model.double()

  # Edge explainability
  # ===================

  # Captum assumes that for all given input tensors, dimension 0 is
  # equal to the number of samples. Therefore, we use unsqueeze(0).
  captum_model = to_captum(model, mask_type='edge', output_idx=output_idx)
  edge_mask    = torch.ones(data.num_edges, requires_grad=True, device=device)

  ig           = IntegratedGradients(captum_model)
  ig_attr_edge = ig.attribute(edge_mask.unsqueeze(0), target=target,
                              additional_forward_args=(data.x, data.edge_index),
                              internal_batch_size=1)

  # Scale attributions to [0, 1]:
  ig_attr_edge = ig_attr_edge.squeeze(0).abs()
  ig_attr_edge /= ig_attr_edge.max()

  # Visualize absolute values of attributions:
  explainer = Explainer(model)
  ax, G = explainer.visualize_subgraph(output_idx, data.edge_index, ig_attr_edge)
  plt.show()

  # Node explainability
  # ===================

  captum_model = to_captum(model, mask_type='node', output_idx=output_idx)

  ig           = IntegratedGradients(captum_model)
  ig_attr_node = ig.attribute(data.x.unsqueeze(0), target=target,
                              additional_forward_args=(data.edge_index),
                              internal_batch_size=1)

  # Scale attributions to [0, 1]:
  ig_attr_node = ig_attr_node.squeeze(0).abs().sum(dim=1)
  ig_attr_node /= ig_attr_node.max()

  # Visualize absolute values of attributions:
  ax, G = explainer.visualize_subgraph(output_idx, data.edge_index, ig_attr_edge,
                                      node_alpha=ig_attr_node)
  plt.show()

  # # Node and edge explainability
  # # ============================

  captum_model = to_captum(model, mask_type='node_and_edge',
                          output_idx=output_idx)

  ig = IntegratedGradients(captum_model)
  ig_attr_node, ig_attr_edge = ig.attribute(
      (data.x.unsqueeze(0), edge_mask.unsqueeze(0)), target=target,
      additional_forward_args=(data.edge_index), internal_batch_size=1)

  # Scale attributions to [0, 1]:
  ig_attr_node = ig_attr_node.squeeze(0).abs().sum(dim=1)
  ig_attr_node /= ig_attr_node.max()
  ig_attr_edge = ig_attr_edge.squeeze(0).abs()
  ig_attr_edge /= ig_attr_edge.max()

  # Visualize absolute values of attributions:
  ax, G = explainer.visualize_subgraph(output_idx, data.edge_index, ig_attr_edge,
                                      node_alpha=ig_attr_node)
  plt.show()
```

running it like this:

```
class GCN(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = GCNConv(num_of_feats, output_dim)
      
    def forward(self, x, edge_index):
        x = F.relu(self.conv1(x, edge_index))
        return x



num_of_nodes = 14
num_of_feats = 5
output_dim   = 10
device       = 'cpu'
adj          = np.random.randint(0,2,[num_of_nodes,num_of_nodes])
feats        = np.random.uniform(-1,1,[num_of_nodes, num_of_feats])

model      = GCN()
captum_explain_graph(model, feats, adj, 1, 'cpu')
```

Issues:

captum requires target in ```ig.attribute``` but I am using GCN as feature extractor so I don't have target values, I have only adj_matrix and feats per node, what should I use for target value in this case? If I use `None` (as per captum), it gives an error. What is the solution for this?

Here is full working code:

[colab link](https://colab.research.google.com/drive/1ijIWA0J4E_w_2A6biYafSt4oLE8sIRBl?usp=sharing)

**Comment by aobo-y** [comment]
hey @monk1337 , `target` is only to simplify the usage of classification model. Captum's IG does not require `target` as long as you return a single value for each input instance in the batch. IG uses the value as the objective/loss to calculate the gradient. You can wrap your model to calculate such objective/loss using the original output of GCN. And pass such wrapper to Captum.

**Comment by craymichael** [comment]
Hi, this is covered in the FAQ target section. Captum attributes to one scalar objective per example; if the model output is not already that shape, wrap the forward function to select or aggregate the output you want to explain. Closing as documented. https://captum.ai/docs/faq
